src/core/shutdown.c

```diff
@@ -158,6 +158,41 @@ static int switch_root_initramfs(void) {
         return switch_root("/run/initramfs", "/oldroot", false, MS_BIND);
 }
 
+static int sync_with_timeout(void) {
+        sigset_t mask, oldmask;
+        pid_t pid;
+        int r;
+
+        log_info("Syncing filesystems.");
+
+        assert_se(sigemptyset(&mask) == 0);
+        assert_se(sigaddset(&mask, SIGCHLD) == 0);
+        assert_se(sigprocmask(SIG_BLOCK, &mask, &oldmask) == 0);
+
+        /* Due to the possiblity of a remount operation hanging, we
+         * fork a child process and set a timeout. If the timeout
+         * lapses, the assumption is that that particular remount
+         * failed. */
+        pid = fork();
+        if (pid < 0)
+                return log_error_errno(errno, "Failed to fork: %m");
+
+        if (pid == 0) {
+                /* Start the mount operation here in the child */
+                sync();
+                _exit(EXIT_SUCCESS);
+        }
+
+        r = wait_for_terminate_with_timeout(pid, LONG_TIMEOUT_USEC);
+        if (r < 0) {
+                log_notice("Timed out, issuing SIGKILL to "PID_FMT".", pid);
+                (void) kill(pid, SIGKILL);
```

Comment:
This kills `pid` and claims a timeout if `pid` _did_ exit, but with something other than `EXIT_SUCCESS`.  Unlikely, I know.  But the existing pattern @poettering had suggested to follow in his review, would avoid this.  So it shouldn't be too hard, and it makes the code easier to understand if you're familiar with the convention of the already existing function.

Reply:
Hmm.  That convention was to return the exit status as a positive (or zero) value, so it doesn't actually help for abnormal exits.  So I think it should instead use `-ETIMEDOUT` to signal timeout specifically.